### PR TITLE
Homework 16 inheritance

### PR DESCRIPTION
**Pull request by nkstlrv**
Had problems with task 4.
Haven't quite understood Custom Exceptions syntax and logic, despite saw tutorials.

### CODE REVIEW

**Comment by nkstlrv** [comment]
Hello! 
There is a new directory, called **(task_3_hw17)**. 

Inside there are two modules:

1) **task_3_original** was made by myself before today's lesson
2) **task_3_new** is the newest one. It is made by taking into account the methods of today's lesson

File: Module 2/HW_16_Inheritance/task_1_hw16.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
+
+class Person:
+
+    def __init__(self, name: str, surname: str, age: int, country: str):
+        
+        self.name = name
+        self.surname = surname
+        self.age = self.age_validator(age)
+        self.age = age
```
Comment (SergeyTovstonog) [inline Module 2/HW_16_Inheritance/task_1_hw16.py:9]: `self.age = age` is not needed

File: Module 2/HW_16_Inheritance/task_1_hw16.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    # Checks whether you passed or failed exam
+    def fail_pass(self, mark):
+        if mark < 60:
+            return "You've failed"
+        else:
+            return "You've passed"
+    # Converts marks from (0-100) system to the (2F - 5A) system  
+    def digit_to_letter_mark_converter(self, mark): #I tried to to this methof with match/case but didn't understand syntax
+        if mark in range(0, 60):
```
Comment (SergeyTovstonog) [inline Module 2/HW_16_Inheritance/task_1_hw16.py:48]: probably like that

```
match mark:
    case mark < 60:
        ...
```

File: Module 2/HW_16_Inheritance/task_3_hw16/task_3_new.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+
+pq_1 = ProductsQuantity(p_1, 0, 1.0)
+
+
+class Market:
+    markup = 1.30
+
+    def __init__(self):
+        self.stock: list[ProductsQuantity] = [pq_1]
```
Comment (SergeyTovstonog) [inline Module 2/HW_16_Inheritance/task_3_hw16/task_3_new.py:36]: `self.stock: list[ProductsQuantity] = []` - there could be empt store.

File: Module 2/HW_16_Inheritance/task_3_hw16/task_3_new.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                break
+            else:
+                self.stock.append(ProductsQuantity(Product(pr_type, name, price), amount, discount))
+                break
+
+    def sell_product(self, name, amount):
+        test_list = []
+        for item in self.stock:
+            if item.product.name == name:
+                test_list.append(item)
```
Comment (SergeyTovstonog) [inline Module 2/HW_16_Inheritance/task_3_hw16/task_3_new.py:57]: no need for list.
`could be item_for_sale = item`

File: Module 2/HW_16_Inheritance/task_4_hw16.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+# New class that inherits previos and logs Exception message to txt file
+class PrintCustomException(CustomException):
+
+    def __init__(self, message="Custom exception occured and printed "):
+        super().__init__(message)
+
+    def raise_exception(self):
+        print_message = str(CustomException(self.message))
+        with open ("Module 2\HW_16_Inheritance\log.txt", "a") as f:
```
Comment (SergeyTovstonog) [inline Module 2/HW_16_Inheritance/task_4_hw16.py:20]: that part with write-to files should be in `CustomException.__init__`

**Review by SergeyTovstonog — APPROVED** [review]
Well done! Kudos for adding tests!

**Pull request merged**